examples/nlp/language_modeling/megatron_change_num_partitions.py

```diff
@@ -148,6 +163,317 @@ def compute_tp_splits(
+    # and semantics of the layer should match.
+    logging.debug(f"Index: {idx} Model Params : {param_name} - {param.shape}")
+    logging.debug(f"Index: {idx} Global params: {partitions[1][idx]} - {partitions[0][idx].shape}")
+
+
+################
+### Handlers ###
+################
+
+
+class GPTHandler:
+    def __init__(self, megatron_legacy: bool):
+        self.duplicate_gpt_word_embedding_offset = 0
+        self.untied_gpt_embedding = False
+        self.megatron_legacy = megatron_legacy
+
+    def compute_split_index(self, model, idx, tp_rank, pp_rank, pp_split_rank, tp_size, pp_size):
+        if pp_rank == (pp_size - 1) and hasattr(model, 'model') and hasattr(model.model, 'word_embeddings'):
+            # duplicate embedding copy (tied weights)
+            self.duplicate_gpt_word_embedding_offset = 1
+
+        if model.cfg.get('share_embeddings_and_output_weights', True) is False:
+            self.untied_gpt_embedding = True
+
+        if self.duplicate_gpt_word_embedding_offset > 0:
+            logging.info(f"GPT duplicate_gpt_word_embedding_offset: {self.duplicate_gpt_word_embedding_offset}")
+
+        return idx + self.duplicate_gpt_word_embedding_offset
+
+    def compute_splits(self, model, partitions, idx, tp_rank, pp_rank, pp_split_rank, tp_size, pp_size):
+        splits = []
+
+        # This is the PP X TP Y model with partial parameters present in correct order.
+        # We need to extract the parameters from the global map in reverse order to fill in the
+        # parameters of this model in forward order.
+        for param_name, param in model.named_parameters():
+
+            # Since we are moving forward, we may reach the end of the global map
+            # but GPT has an additional word embedding as its last parameter
+            # Therefore we check for this, and reset the index to the parameter of the PP 0 TP 0 rank
+            # which holds the parameters of the embedding.
+            if idx == (len(partitions[0])) and self.duplicate_gpt_word_embedding_offset > 0:
+                logging.info("Found duplicate embedding copy for GPT model, resetting index")
+                idx = 0  # reset idx parameter to 0 if we have duplicate embedding copy
+
+            debug_log_split_param_diff(idx, param, param_name, partitions)
+
+            # Tensor Parallel Splitting
+            split = compute_tp_splits(
+                param_name, param, partitions, idx, tp_size, pp_size, pp_rank, pp_split_rank, self.megatron_legacy
+            )
+
+            splits.append(split)
+            idx += 1
+
+        return idx, splits
+
+    def compute_split_offset(self, offset_diff, tp_rank, pp_rank, pp_split_rank, tp_size, pp_size):
+        # GPT offset correction
+        if not self.untied_gpt_embedding and pp_size > 1 and pp_rank == (pp_size - 1) and pp_split_rank == 0:
+            offset_diff += 1
+
+        return offset_diff
+
+
+class T5Handler:
```

Comment:
with this handler way of doing things, how do we handle other models? like say BERT? do we add a handler for each model?

Reply:
Yep, we'll need a handler for each model. Every minor change in the model's arch can shift around too many layers or require duplication which complicates Pipeline parallelism merging.